Complete review of the "Customize Error Pages" cookbook article

javiereguiluz opened the pull request:
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | all |
| Fixed tickets | - |

In my opinion the current article looks "old" compared to other parts of the documentation. I've reworded it completely to make it more actionable and to explain more clearly the three alternatives provided to developers.

javiereguiluz committed:
Complete review of the "Customize Error Pages" cookbook article

javiereguiluz committed:
Fixed a reStructuredText formatting issue

javiereguiluz committed:
Fixed a restructuredtext syntax issue

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,154 @@
 
-Thus, error pages can be customized in different ways, depending on how
-much control you need:
+.. image:: /images/cookbook/controller/error_pages/exceptions-in-dev-environment.png
```
You have to remove the `controller/` part of the path.

javiereguiluz replied:
Thanks. Actually I think the error was that images should be under `controller/error_pages/`. I've moved the image files.

aitboudad commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,154 @@
+recommended because it will also stop Symfony from recompiling your Twig templates,
+among many other things.
+
+The recommended solution is to use a third-party bundle called `WebfactoryExceptionsBundle`_.
```
Do not forgot to change this part for >=2.6 as the main functionality of WebfactoryExceptionsBundle has been integrated into Symfony core.

javiereguiluz replied:
:+1: You are right. This article has changed a lot in 2.6.

javiereguiluz committed:
Fixed the location of the cookbook article images

WouterJ commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,154 @@
-a configurable (but otherwise arbitrary) controller to generate the
-response. The default controller used has a sensible way of
-picking one out of the available set of error templates.
+In the `development environment`_ Symfony catches all the exceptions and displays
```
"In the development environment, Symfony catches [...]"

WouterJ commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,154 @@
-   templates that allow you to customize how your different error
-   pages look (easy); <use-default-exception-controller>`
+Since these pages contain a lot of sensitive internal information about your
+application, in production environment Symfony displays instead a simple and
```
This is a very big sentence, let's break it up: "Since these pages contain a lot of sensitive internal information, Symfony won't display this in the production environment. Instead, it'll show a simple and generic **error page**."

javiereguiluz replied:
I love short sentences! Done. Thanks.

javiereguiluz committed:
Rewordings

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,154 @@
-Thus, error pages can be customized in different ways, depending on how
-much control you need:
+.. image:: /images/cookbook/controller/error_pages/exceptions-in-dev-environment.png
+   :alt: A typical exception page in development environment
```
[...] in the development [...]

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,154 @@
-#. :ref:`Replace the default exception controller with your own
-   (intermediate). <custom-exception-controller>`
+.. image:: /images/cookbook/controller/error_pages/errors-in-prod-environment.png
+   :alt: A typical error page in production environment
```
[...] in the production [...]

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,154 @@
 
-#. :ref:`Use the kernel.exception event to come up with your own
-   handling (advanced). <use-kernel-exception-event>`
+Error pages for production environment can be customized in different ways,
```
[...] for the production [...]

xabbuh replied:
The comma here looks wrong to me.

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,154 @@
+Then, the controller selects one of the templates defined in the
+``Resources/views/Exception`` directory of the TwigBundle to render the error
+page. If you browse that directory (usually located in
+``vendor/symfony/symfony/src/Symfony/Bundle/TwigBundle``) you'll find a lot of
```
What do you think about linking to it on GitHub instead?

javiereguiluz replied:
I'm not sure about this.

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,154 @@
-How the Template for the Error and Exception Pages Is Selected
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The logic followed by the ``ExceptionController`` to pick one of the available
+templates is based on the HTTP status code and request format:
```
[...] and the request [...]

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,154 @@
-    error page for many of the most common response formats, including
-    JSON (``error.json.twig``), XML (``error.xml.twig``) and even
-    JavaScript (``error.js.twig``), to name a few.
+#. If none of the previous template exist, fall back to the generic HTML template
```
templates

javiereguiluz replied:
I think it should be singular because in this case there is only one template.

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,154 @@
+``error.html.twig`` template to catch any other error different from 404 and 500.
 
-Overriding or Adding Templates
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
A label for the old headline should be added before the paragraph starting in line 68.

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,154 @@
-.. note::
+One of the biggest hurdles of testing how do custom error pages look in your
+application is the fact that Symfony ignores them in the development environment
+and displays instead the default exception pages.
```
and displays the default exception pages instead

or

and instead displays the default exception pages

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,154 @@
 .. _custom-exception-controller:
 
-Replacing the Default ExceptionController
+Overriding the Default ExceptionController
```
We need a label for the old headline.

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -205,86 +187,65 @@ to point to it.
-    described previously, which is registered in the DIC as the
-    ``twig.controller.exception`` service.
+The :class:`Symfony\\Component\\HttpKernel\\EventListener\\ExceptionListener`
+class used by TwigBundle as a listener of the ``kernel.exception`` event creates
```
[...] by the TwigBundle [...]

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -205,86 +187,65 @@ to point to it.
-    ``twig.controller.exception`` service.
+The :class:`Symfony\\Component\\HttpKernel\\EventListener\\ExceptionListener`
+class used by TwigBundle as a listener of the ``kernel.exception`` event creates
+the Request that will be dispatched to your controller. In addition, your controller
```
I think we should enclose "Request" with literals if we want to refer to the class. Otherwise, it should be lowercased.

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -205,86 +187,65 @@ to point to it.
-    This event listener is set up by the TwigBundle.
+``logger``
+    A :class:`\\Symfony\\Component\\HttpKernel\\Log\\DebugLoggerInterface`
+    instance which may be ``null`` in some circumstances.
```
I think it should be explained when and why it can be `null`.

javiereguiluz replied:
I need help with this because I don't know when or why :(

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -205,86 +187,65 @@ to point to it.
-``ExceptionController`` could do for you was trying to figure out the
-right HTTP status code and display your nice-looking error page.
+Working with this event is actually much more powerful than what has been explained
+before but also requires a thorough understanding of Symfony internals. Suppose
```
[...] before, but [...]

xabbuh commented on cookbook/controller/error_pages.rst:
```diff
@@ -205,86 +187,65 @@ to point to it.
+Working with this event is actually much more powerful than what has been explained
+before but also requires a thorough understanding of Symfony internals. Suppose
+that your code throws specialized exceptions with a particular meaning to your
+application domain.
```
I would move the last sentence at the beginning of the following paragraph.

javiereguiluz commented:
I've added all the suggested changes. @xabbuh please review that *"the labels for the old and removed headings"* are OK because I'm not sure what I'm doing.

javiereguiluz committed:
Changes suggested by reviewers

WouterJ commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,157 @@
-#. :ref:`Replace the default exception controller with your own
-   (intermediate). <custom-exception-controller>`
+.. image:: /images/cookbook/controller/error_pages/errors-in-prod-environment.png
+   :alt: A typical error page in the production environment
```
maybe add a tip that you can see the production error page in the dev env by clicking on the link?

weaverryan replied:
what link can you click to see the production error page?

WouterJ replied:
I thought you added that feature?

weaverryan replied:
I had a PR for that, but then my PR was replaced with a whole other, better proposal for 2.6 (which I don't believe has any link like that).

WouterJ commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,157 @@
+depending on your needs:
+
+#. If you just want to change the contents and styles of the error pages to match
+   the rest of your application, :ref:`override default error templates <use-default-exception-controller>`.
```
all list items should end with `;`, except from the last one (which should end with `.`)

WouterJ commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,157 @@
 
-Using the Default ExceptionController
--------------------------------------
+Overriding the Default Error Templates
```
you should add `.. _using-the-default-exceptioncontroller:`

WouterJ commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,157 @@
-By default, the ``showAction()`` method of the
-:class:`Symfony\\Bundle\\TwigBundle\\Controller\\ExceptionController`
-will be called when an exception occurs.
+By default, when an exception occurs, the ``showAction()`` method of the
```
I think it's better to say "By default, the `showAction()` method of the `ExceptionController` is called whenever an exception occurs, thanks to an event listener configured by the TwigBundle."

WouterJ commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,157 @@
+Then, the controller selects one of the templates defined in the
+``Resources/views/Exception`` directory of the TwigBundle to render the error
+page. If you browse that directory (usually located in
+``vendor/symfony/symfony/src/Symfony/Bundle/TwigBundle``) you'll find a lot of
```
maybe link to github?

WouterJ commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,157 @@
+``Resources/views/Exception`` directory of the TwigBundle to render the error
+page. If you browse that directory (usually located in
+``vendor/symfony/symfony/src/Symfony/Bundle/TwigBundle``) you'll find a lot of
+templates defined for different types of errors and content formats.
```
maybe also mention that `error.*.twig` is used in prod and `exception.*.twig` in dev? (if I remember correctly)

WouterJ commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,157 @@
-    called to obtain the HTTP status code to use. Otherwise,
-    the status code will be "500".
+    the ``status_code`` variable will contain the value returned by the
+    ``getStatusCode()`` method. Otherwise, the ``status_code`` variable will be ``500``.
```
this sentence (if I can read it correctly due to the crazy diffs) is not very great to read. What about:

```rst
You can customize the status code by implementing
:class:`Symfony\\Component\\HttpKernel\\Exception\\HttpExceptionInterface`
and its required ``getStatusCode()`` method. Otherwise, the ``status_code`` will default
to ``500``.
```

WouterJ commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,157 @@
+    The exception pages shown in the development environment can be customized
+    in the same way as error pages. Create a new ``exception.html.twig`` template
+    for the standard HTML exception page or ``exception.json.twig`` for the JSON
+    exception page.
```
ah, here it is! I propose to move this somewhat above

WouterJ commented on cookbook/controller/error_pages.rst:
```diff
@@ -5,172 +5,157 @@
-For example, to override the default error template, create a new
-template located at
-``app/Resources/TwigBundle/views/Exception/error.html.twig``:
+Avoiding Exceptions when Using Security Functions in Error Templates
```
:+1: (maybe a sidebar is better though?)

javiereguiluz committed:
Address most of the comments made by Wouter

javiereguiluz commented:
@WouterJ I've addressed most of your comments and suggestions. Could we please label this important PR as "finished" and merge it? Thanks!

weaverryan commented:
Thanks Javier!

Sorry for the slow merge - this kind of update is SO great. I merged it and made some more tweaks and shortenings in #5304.

Cheers!
